sentry_sdk/client.py

```diff
@@ -459,11 +459,11 @@ def _should_sample_error(
         # type: (...) -> bool
         sampler = self.options.get("error_sampler", None)
 
+        sample_rate = self.options["sample_rate"]
```

Comment:
I would prefer to find a slightly different way to solve this bug, since with this change, we will be handling the case where the `error_sampler` returns an invalid value differently from the case where it raises an error, even though it would make the most sense to handle both cases identically.

Specifically, with this change, returning an invalid value from the sampler triggers the `except TypeError` block on line 470, where we log a warning message informing the user of the invalid return value, and then we default to always sampling the event. However, if the sampler function raises an error, we fall back to the `sample_rate` option since the variable assignment from line 462 is never overwritten. If the user has set a custom `sample_rate`, the event might not end up being sampled.

I think we should either (1) default to always sampling when the `error_sampler` has an invalid value (i.e., raises an error, returns invalid value, or is not callable) or (2) fall back to the `sample_rate` and then default to always sampling only if the `sample_rate` is also invalid.

I would slightly prefer option (1), since it also matches the [documented behavior](https://docs.sentry.io/platforms/python/configuration/options/#error-sampler) that the `sample_rate` is ignored when an `error_sampler` is provided. If we instead pick option (2), we would also need to tweak the docs to state that we ignore the `sample_rate` when a _valid_ `error_sampler` is provided.

To implement option (1), I think we could first undo this change completely, and then add an `except NameError` block to handle the case where the `sample_rate` does not get set because the sampler raised an error. Lastly, to ensure a consistent handling when the user has set the `error_sampler` to something that is not a function, we should change the `if callable(sampler)` line to `if sampler is not None`; then, we will end up handling the case of of the user supplying a non-callable `error_sampler` in the same `except NameError` block that also handles an `error_sampler` that raises an error.

Reply:
makes sense. will change this to option (1). Great review @szokeasaurusrex

Reply:
@szokeasaurusrex updated my code. have you had something like this is mind? Please re-review.